### PR TITLE
[ML] Manually set RPATH linker flags for MacOS cross compilation builds

### PR DESCRIPTION
**Pull request by edsavage**
When cross compiling to MacOS CMake ignores the MACOSX_RPATH property
for unknown reasons. This is a workaround to set the appropriate linker
flags manually.

### CODE REVIEW

File: 3rd_party/3rd_party.cmake [hunk 1]
```diff
@@ -1,274 +1,282 @@
...
+  set(HOST_SYSTEM_PROCESSOR "x86_64")
+else()
+  execute_process(COMMAND uname -m OUTPUT_VARIABLE HOST_SYSTEM_PROCESSOR OUTPUT_STRIP_TRAILING_WHITESPACE)
+  string(REPLACE arm aarch HOST_SYSTEM_PROCESSOR ${HOST_SYSTEM_PROCESSOR})
+endif()
+message(STATUS "3rd_party: HOST_SYSTEM_PROCESSOR=${HOST_SYSTEM_PROCESSOR}")
+set(CMAKE_TOOLCHAIN_FILE "cmake/${HOST_SYSTEM_NAME}-${HOST_SYSTEM_PROCESSOR}.cmake")
+
+set(ARCH ${HOST_SYSTEM_PROCESSOR})
+set(CMAKE_SYSTEM_NAME ${HOST_SYSTEM_NAME})
```
Comment (droberts195) [inline 3rd_party/3rd_party.cmake:38]: I still don't understand why we have to redefine a built in CMake variable to have a different value to what it usually has.  Wouldn't it be clearer just to use `${HOST_SYSTEM_NAME}` in the code below that checks the lowercase value?

```suggestion
```

File: 3rd_party/3rd_party.cmake [hunk 2]
```diff
@@ -1,274 +1,282 @@
...
+  set(BOOST_LIBRARIES "atomic" "chrono" "date_time" "filesystem" "iostreams" "log" "log_setup" "program_options" "regex" "system" "thread" "unit_test_framework")
+  set(XML_LOCATION)
+  set(GCC_RT_LOCATION)
+  set(STL_LOCATION)
+  set(OMP_LOCATION)
+  set(ZLIB_LOCATION)
+  set(TORCH_LIBRARIES "torch_cpu" "c10")
+  set(TORCH_LOCATION "/usr/local/lib")
+  set(TORCH_EXTENSION ".dylib")
+elseif ("${CMAKE_SYSTEM_NAME}" STREQUAL "linux")
```
Comment (droberts195) [inline 3rd_party/3rd_party.cmake:60]:
```suggestion
elseif ("${HOST_SYSTEM_NAME}" STREQUAL "linux")
```

File: 3rd_party/3rd_party.cmake [hunk 3]
```diff
@@ -1,274 +1,282 @@
...
+  execute_process(COMMAND uname -m OUTPUT_VARIABLE HOST_SYSTEM_PROCESSOR OUTPUT_STRIP_TRAILING_WHITESPACE)
+  string(REPLACE arm aarch HOST_SYSTEM_PROCESSOR ${HOST_SYSTEM_PROCESSOR})
+endif()
+message(STATUS "3rd_party: HOST_SYSTEM_PROCESSOR=${HOST_SYSTEM_PROCESSOR}")
+set(CMAKE_TOOLCHAIN_FILE "cmake/${HOST_SYSTEM_NAME}-${HOST_SYSTEM_PROCESSOR}.cmake")
+
+set(ARCH ${HOST_SYSTEM_PROCESSOR})
+set(CMAKE_SYSTEM_NAME ${HOST_SYSTEM_NAME})
+
+if ("${CMAKE_SYSTEM_NAME}" STREQUAL "darwin")
```
Comment (droberts195) [inline 3rd_party/3rd_party.cmake:40]:
```suggestion
if ("${HOST_SYSTEM_NAME}" STREQUAL "darwin")
```

File: 3rd_party/3rd_party.cmake [hunk 4]
```diff
@@ -1,274 +1,282 @@
...
+  install_libs("libxml2" ${XML_LOCATION} "" "${XML_EXTENSION}" "libxml2")
+  install_libs("gcc runtime library" ${GCC_RT_LOCATION} "" "${GCC_RT_EXTENSION}" "libgcc_s")
+  install_libs("C++ standard library" ${STL_LOCATION} "" "${STL_EXTENSION}" "${STL_PATTERN}")
+  install_libs("OMP runtime library" ${OMP_LOCATION} "" "${OMP_EXTENSION}" "${OMP_PATTERN}")
+  install_libs("zlib" ${ZLIB_LOCATION} "" "${ZLIB_EXTENSION}" "zlib")
+  install_libs("Torch libraries" ${TORCH_LOCATION} "" "${TORCH_EXTENSION}" "${TORCH_LIBRARIES}")
+  install_libs("Intel MKL libraries" ${MKL_LOCATION} "${MKL_PREFIX}" "${MKL_EXTENSION}" "${MKL_LIBRARIES}")
+
+  # Special extra platform-specific processing
+  if ("${CMAKE_SYSTEM_NAME}" STREQUAL "linux")
```
Comment (droberts195) [inline 3rd_party/3rd_party.cmake:260]:
```suggestion
  if ("${CMAKE_SYSTEM_NAME}" STREQUAL "Linux")
```

File: 3rd_party/3rd_party.cmake [hunk 5]
```diff
@@ -1,274 +1,282 @@
...
+endif()
+message(STATUS "3rd_party: HOST_SYSTEM_PROCESSOR=${HOST_SYSTEM_PROCESSOR}")
+set(CMAKE_TOOLCHAIN_FILE "cmake/${HOST_SYSTEM_NAME}-${HOST_SYSTEM_PROCESSOR}.cmake")
+
+set(ARCH ${HOST_SYSTEM_PROCESSOR})
+set(CMAKE_SYSTEM_NAME ${HOST_SYSTEM_NAME})
+
+if ("${CMAKE_SYSTEM_NAME}" STREQUAL "darwin")
+  message(STATUS "3rd_party: CMAKE_SYSTEM_NAME = ${CMAKE_SYSTEM_NAME}")
+  message(STATUS "3rd_party: Copying macOS 3rd party libraries")
```
Comment (droberts195) [inline 3rd_party/3rd_party.cmake:42]: nit: in all the other branches these messages are the other way around

```suggestion
  message(STATUS "3rd_party: Copying macOS 3rd party libraries")
  message(STATUS "3rd_party: CMAKE_SYSTEM_NAME = ${CMAKE_SYSTEM_NAME}")
```

File: 3rd_party/3rd_party.cmake [hunk 6]
```diff
@@ -1,274 +1,272 @@
...
+  install_libs("gcc runtime library" ${GCC_RT_LOCATION} "" "${GCC_RT_EXTENSION}" "libgcc_s")
+  install_libs("C++ standard library" ${STL_LOCATION} "" "${STL_EXTENSION}" "${STL_PATTERN}")
+  install_libs("OMP runtime library" ${OMP_LOCATION} "" "${OMP_EXTENSION}" "${OMP_PATTERN}")
+  install_libs("zlib" ${ZLIB_LOCATION} "" "${ZLIB_EXTENSION}" "zlib")
+  install_libs("Torch libraries" ${TORCH_LOCATION} "" "${TORCH_EXTENSION}" "${TORCH_LIBRARIES}")
+  install_libs("Intel MKL libraries" ${MKL_LOCATION} "${MKL_PREFIX}" "${MKL_EXTENSION}" "${MKL_LIBRARIES}")
+
+  # Special extra platform-specific processing
+  if ("${HOST_SYSTEM_NAME}" STREQUAL "linux")
+    if (INSTALL_DIR AND NOT DEFINED ENV{CPP_CROSS_COMPILE} OR NOT "$ENV{CPP_CROSS_COMPILE}" STREQUAL "macosx")
```
Comment (droberts195) [inline 3rd_party/3rd_party.cmake:251]: I think given the code above these two conditions together could now be simplified to:

```
if (INSTALL_DIR AND CMAKE_SYSTEM_NAME STREQUAL "Linux")
```

**Review by droberts195 — APPROVED** [review]
LGTM

I saw a potential simplification but it's not important so happy to merge as-is if you prefer

File: 3rd_party/3rd_party.cmake (hunk 6 above)
Comment (edsavage) [reply 3rd_party/3rd_party.cmake:251]: Thanks Dave. At the moment it seems to be working so I think I'll leave well enough alone!

**Pull request merged**